### ISSUE TITLE
[db_partition] job event migration fails due to missing index

### ISSUE THREAD

**Issue by jladdjr** [ansible/awx#9520]
* The current job event migration was based on the previous bigint migration.
* One notable change was made in the job event migration, though - the `id` primary key was dropped (postgres requires all uniqueness constraints to include the partition key)
* When an upgrade is performed w/ a database filled with billions of job events, the migration stalls because of the missing index

**Comment by jladdjr** [comment]
@Ladas provided some thoughts for how to address this here:
https://github.com/ansible/awx/pull/9341#discussion_r589687501

**Comment by jladdjr** [comment]
Grabbed bits of info from off-post chat:

```
@jladd you're still ordering by ID in the migration, so that will cause sequential scan

@jladd you're removing the primary key because of the create table like right? I'd just create a temp table with create table like job_events , then remove the primary constraint from this temp table and do the parent of partitions from it :slightly_smiling_face: (so we don't need to touch the original table)

@jladd @ryanpetrello btw. you can insert+delete the batch in 1 query I think, using CTE
WITH tmp AS (DELETE FROM table RETURNING column1, column2, ...)
INSERT INTO another_table (column1, column2, ...) SELECT * FROM tmp;


and to get amount of inserted cols (to know when to end), this should work:
con = engine.raw_connection()
cursor = con.cursor()
cursor.execute(query.as_string(cursor))
con.commit()
count = cursor.rowcount # amount of inserted/updated/deleted rows

instead of using MAX(id)
11:03
just do SELECT ... FROM LIMIT 1


@jladd I'd say copy original schema to TMP table -> change TMP table schema(remove the primary key) -> create partition parent from TMP table schema  is probably the simplest (one line change, or 2 lines, deleting also the TMP table :slightly_smiling_face: ) (edited) 

 @jladd each of the MAX(id), INSERT and DELETE queries is causing sequential scan now, so I think the index on id is needed, if you want to use the current way (although there should be a way to move the data with DB procedure without index, in one sequential pass of the table, using a DB cursor :slightly_smiling_face: , but it's better to stay with working algorithm as @ryanpetrello says :slightly_smiling_face: )
11:23
I'll move this comment to GH, so it doesn't get lost :slightly_smiling_face:
```

**Comment by jladdjr** [comment]
https://github.com/jladdjr/awx/commit/91b08176884690ef6f3ec69a3cfada884f597571 creates a tmp table that we modify / use as a basis for the partitioned table's schema

**Comment by jladdjr** [comment]
Okay, have some results on the migration tweaks:
* Spun up a 3.8 instance
* Ran jobs every minute, where each job created ~6k job events, until I had a total of 1.2 million job events. Not huge, but was enough to actually exercise the migration in a non-trivial way

The results looked really good:
- migrations finished in a matter of minutes
- verified all job events made their way to the new partition table (main_jobevent_old_events;)
- verified I can render the job events in tower

**Comment by jladdjr** [comment]
Plus, the metrics for retrieving api/v2/jobs and api/v2/job/N/job_events are looking good (this may not mean much given this is only 1.2M job events, but at least it shows that at this number things are snappy, and hopefully with 1B events they stay that way):

Getting job events for one job - `0.041s`:
![image](https://user-images.githubusercontent.com/4440360/110436395-08a6c280-8069-11eb-869c-e0655965ea79.png)

Getting job list itself - `0.153s`:
![image](https://user-images.githubusercontent.com/4440360/110436416-10fefd80-8069-11eb-9e13-b0d7d62208a5.png)

In short, I think we have our next candidate to test. I think we should try to monkey patch the database if we can.
* Reinstate the id index for the old job events table (this may take a long time)
* .. kick the task manager and see if the workers have an easier time w/ the migration

@ryanpetrello are we wanting to try to get rid of this (in favor of the LIMIT 1 query, or something else)?
https://github.com/jladdjr/awx/blob/18ba40506f2d1d4f38b78e2479cdf78d471544ec/awx/main/tasks.py#L758

**Comment by jladdjr** [comment]
That was an m5.large instance:
2 CPU, 8 GB Memory
Not sure what the iops are, but guessing that - like the other specs - they’re modest

**Comment by jladdjr** [comment]
Going to try testing these changes at scale today - to begin with, will need to re-introduce the main_jobevent id index that was dropped. That will take a while (hopefully no more than a day). To speed things up, will try giving the background tasks as many resources as we can.

**Comment by jladdjr** [comment]
cc @ryanpetrello @kdelee @mcharanrm ^

**Comment by jladdjr** [comment]
Action items:

- [x] bump `maintenance_work_mem` on the PG server config to couple of GBs
- [x] `max_parallel_maintenance_workers` - we think PG12 creates index in parallel, so more workers, the better (but workers <= # cores)
- [x] we can do `maintenance_work_mem * max_parallel_maintenance_workers < 70-80% MAX mem`, then revert it back after index creation

**Comment by jladdjr** [comment]
Given these machine specs:
* 8 cores
* 64 GB memory

Setting:
* maintenance_work_mem = 7GB
* max_parallel_maintenance_workers = 8

So that `maintenance_work_mem` * `max_parallel_maintenance_workers` = 56GB < 64GB available memory (leaving some headroom)

config file:
/var/lib/pgsql/data/postgresql.conf

---

Background info on settings:

maintenance_work_mem
* https://www.postgresql.org/docs/12/runtime-config-resource.html

max_parallel_maintenance_workers
* https://www.postgresql.org/docs/12/runtime-config-resource.html#GUC-MAX-PARALLEL-WORKERS-MAINTENANCE

**Comment by jladdjr** [comment]
Reloading config:
```
su - postgres
pg_ctl reload
```

**Comment by jladdjr** [comment]
Running:
`ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id);`

^ threw that in a screen session since I'll almost certainly have my connection to the VPN reset while this runs

**Comment by jladdjr** [comment]
`top` doesn't seem to show as much activity as I would have expected on the machine.

Running the `show all;` pg command:
* confirms that `maintenance_work_mem` is set to `7GB`
* .. but I'm not seeing `max_parallel_maintenance_workers` listed anywhere

`max_parallel_maintenance_workers = 8` is definitely in /var/lib/pgsql/data/postgresql.conf

looks like if I ask what that setting's value is explicitly, then it will confirm the setting took effect:

```
show max_parallel_maintenance_workers;
 max_parallel_maintenance_workers
----------------------------------
 8
```

.. so maybe we're fine. Still, wish I had a way to confirm indexing was moving along at the pace we expect.

@ryanpetrello @Ladas - any tips for how to do this? ^

**Comment by jladdjr** [comment]
I can at least confirm that postgres is busy working on the index:

```
awx=> SELECT
  pid,
  now() - pg_stat_activity.query_start AS duration,
  query,
  state
FROM pg_stat_activity
WHERE (now() - pg_stat_activity.query_start) > interval '5 minutes' and state = 'active';
   pid   |    duration    |                                               query                                                | state
---------+----------------+----------------------------------------------------------------------------------------------------+--------
 3966056 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966016 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966057 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966058 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966059 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966060 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966061 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
 3966062 | 00:33:54.51339 | ALTER TABLE ONLY _unpartitioned_main_jobevent ADD CONSTRAINT main_jobevent_pkey1 PRIMARY KEY (id); | active
```

**Comment by jladdjr** [comment]
https://www.postgresql.org/docs/12/progress-reporting.html

```
awx=> select datname, command, phase, blocks_total, blocks_done, div(100 * blocks_done, blocks_total) as percentage_done from pg_stat_progress_create_index;
 datname |   command    |             phase              | blocks_total | blocks_done | percentage_done
---------+--------------+--------------------------------+--------------+-------------+-----------------
 awx     | CREATE INDEX | building index: scanning table |    353676685 |   105363261 |              29
```

About a third of the way done after 2 hours. So we've got about four more to go (assuming this stage represents the main index creation stage [1]). Should be done around 7pm PT / 10pm PT.

[1] Slide 12 of [this deck](https://wiki.postgresql.org/images/3/38/Progress-report.pdf) suggests that it probably is the last major step in the index creation:

![image](https://user-images.githubusercontent.com/4440360/110551480-72b37c00-80ea-11eb-917d-b2bc376f966c.png)

**Comment by jladdjr** [comment]
58% done

```
awx=> select datname, command, phase, blocks_total, blocks_done, div(100 * blocks_done, blocks_total) as percentage_done from pg_stat_progress_create_index;
 datname |   command    |             phase              | blocks_total | blocks_done | percentage_done
---------+--------------+--------------------------------+--------------+-------------+-----------------
 awx     | CREATE INDEX | building index: scanning table |    353676685 |   205800452 |              58
(1 row)
```

**Comment by jladdjr** [comment]
6.5 hours, 95%

```
awx=> select datname, command, phase, blocks_total, blocks_done, div(100 * blocks_done, blocks_total) as percentage_done from pg_stat_progress_create_index;
 datname |   command    |             phase              | blocks_total | blocks_done | percentage_done
---------+--------------+--------------------------------+--------------+-------------+-----------------
 awx     | CREATE INDEX | building index: scanning table |    353676685 |   337304254 |              95
```

**Comment by jladdjr** [comment]
We have an index \o/

![image](https://user-images.githubusercontent.com/4440360/110579222-2d5b7280-811b-11eb-9946-5c2abd463893.png)

**Comment by jladdjr** [comment]
Restored old postgres.conf file (to restore memory / maintenance worker settings):
`[postgres@gprfs047 ~]$ pg_ctl reload`

**Comment by jladdjr** [comment]
Restarted tower services, confirmed services were running:
```
ansible -i inventory tower -m shell -a 'ansible-tower-service start'
ansible -i inventory tower -m shell -a 'ansible-tower-service status'
```

**Comment by jladdjr** [comment]
seeing first report of rows being migrate:
```
2021-03-10 05:14:16,256 WARNING  [-] awx.main.tasks migrated rows to partitioned main_jobevent from _unpartitioned_main_jobevent; # (1775685315 rows remaining)
```

**Comment by jladdjr** [comment]
postgres is showing the migration query:
```
awx=> SELECT
  pid,
  now() - pg_stat_activity.query_start AS duration,
  query,
  state
FROM pg_stat_activity
WHERE (now() - pg_stat_activity.query_start) > interval '1 minutes' and state = 'active';
   pid   |    duration     |                                                                     query                                                                      | state
---------+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------+--------
 3978935 | 00:01:43.646759 | INSERT INTO main_jobevent SELECT *, '1970-01-01' as job_created FROM _unpartitioned_main_jobevent ORDER BY id DESC LIMIT 1000000 RETURNING id; | active
```

Looks like we're moving 1,000,000 rows at a time.

**Comment by jladdjr** [comment]
Seeing second log line for migration. This means that the first migration setup completed!
```
2021-03-10 05:14:16,256 WARNING  [-] awx.main.tasks migrated rows to partitioned main_jobevent from _unpartitioned_main_jobevent; # (1775685315 rows remaining)
2021-03-10 05:17:36,534 WARNING  [-] awx.main.tasks migrated rows to partitioned main_jobevent from _unpartitioned_main_jobevent; # (1774685315 rows remaining)
```

Estimated how long this will take based on those two data points (just to get a ballpark sense for how long this will run):

1775685315 total rows * (200.278 seconds / 10^6 rows) = 355,630 seconds
.. which is 98.8 hours
.. which is 4 days 3 hours

So this should end at Sunday 3am ET.

@Ladas @ryanpetrello - the above back-of-the-envelope calculation assumes that each million rows will take the same amount of time to migrate. Should we expect the operation to take less time as the number of remaining records gets smaller? Will postgres be vacuuming / updating the `id` index in a way that will make the INSERT statement happen more quickly?

@ryanpetrello - how does this performance compare to the bigint migration?

**Comment by jladdjr** [comment]
Job Details page from a page run before the migration:
![image](https://user-images.githubusercontent.com/4440360/110581426-58e05c00-811f-11eb-90ec-09e55520919e.png)

Corresponding /api/v2/jobs/17427/job_events page loads in `0.5 seconds`:
![image](https://user-images.githubusercontent.com/4440360/110581458-731a3a00-811f-11eb-833a-aa3e36d8fde2.png)

Jobs page is loading in `1.5 seconds`:
![image](https://user-images.githubusercontent.com/4440360/110581514-89c09100-811f-11eb-8beb-4898bebd08e0.png)

Those pages are loading much more quickly (based on my memory of the old numbers), _while the migration is in place and the database is busy churning away_.

**Comment by jladdjr** [comment]
Ran a fresh job and it was successful!

![image](https://user-images.githubusercontent.com/4440360/110583275-4e739180-8122-11eb-8dfd-02dd62af28a1.png)

.. and the job events get loaded!

![image](https://user-images.githubusercontent.com/4440360/110583350-68ad6f80-8122-11eb-9f45-36fa331af050.png)

**Comment by jladdjr** [comment]
.. and look, a new partition!

![image](https://user-images.githubusercontent.com/4440360/110583425-867ad480-8122-11eb-94ea-bfc1adfecd79.png)

with just the right number of events for the job I just ran!

```
awx=> select count(*) from main_jobevent_20210310_05;
 count
--------
 102004
(1 row)
```

**Comment by jladdjr** [comment]
Got to call it a night, but this is all looking really good as far as I can tell.

Closing.

**Comment by jladdjr** [comment]
Updated estimate of how long migration will take base on 45 minutes of migrations - looks like it will take a few more hours. That may be because I've been running jobs and poking around.

1775685315 total rows * (2742 seconds / 13*10^6 rows) = 355,630 seconds
.. which is 104 hours (up from 98.8 hours)
